src/cache-worker.ts

```diff
@@ -63,35 +82,65 @@ export class CacheWorker {
     this.pagination = pagination
   }
 
+  /**
+   * Requests Serlo's API to update its cache according to chosen keys.
+   * @param keys an array of keys(strings) whose values should to be cached
+   */
   public async update(keys: string[]): Promise<void> {
-    const keysBlocks = this.splitUpKeysIntoChunks(keys, this.pagination)
-    await this.requestUpdateByBlocksOfKeys(keysBlocks)
+    if (keys.length === 0) {
+      throw new Error('EmptyCacheKeysError: no cache key was provided')
+    }
+    this.tasks = this.makeStackOutOfKeys(keys, this.pagination)
+    await this.makeRequests()
   }
 
-  // TODO: change return type to queue
-  private splitUpKeysIntoChunks(
-    keys: string[],
-    pagination: number
-  ): string[][] {
-    const keysClone = [...keys]
-    const chunksOfKeys: string[][] = []
-    while (keysClone.length) {
-      const temp = keysClone.splice(0, pagination)
-      chunksOfKeys.push(temp)
+  private makeStackOutOfKeys(keys: string[], pagination: number): Stack<Task> {
+    const chunksOfKeys = splitEvery(pagination, keys)
+    const tasks = new Stack<Task>()
+    chunksOfKeys.forEach((chunk) => {
+      const task = { keys: chunk, numberOfRetries: 0 }
+      tasks.push(task)
+    })
+    return tasks
+  }
+
+  private async makeRequests() {
+    while (!this.tasks.isEmpty()) {
+      const task = this.tasks.peekAndPop()
+      const { response, hasError } = await this.getResponse(task)
+      if (hasError && task.keys.length > 1) {
+        this.bisect(task)
+      } else if (hasError && task.keys.length === 1) {
+        await this.retry(task)
+      } else {
+        this.fillLogs(response)
+      }
```

Comment:
```suggestion
      const task = this.tasks.pop()
      const { response, hasError } = await this.getResponse(task)
      if (hasError) {
        if (task.keys.length > 1) {
          this.bisect(task)
        } else if (task.numberOfRetries < MAX_RETRIES) {
          task.push({ ...task, numberOfRetries: task.numberOfRetries + 1})
        } else {
          this.errorLog.push(response)
        }
      } else {
        this.okLog.push(response)
      }
```

Thus, you do not need `fillLogs()` and `retry()`.

Reply:
You might need https://github.com/serlo/api.serlo.org-cache-worker/pull/8#discussion_r530517647 as well for this implementation

Reply:
It would be an very elegant refactoring.
But for now I feel more comfortable maintaining the functions, for simple responsibility and readability.

Reply:
you mean you would do this as a follow up task?

Reply:
That could be a subject of a follow-up issue, yes. We could discuss it there